### PR TITLE
Remove usages of assert in all code that ends up bundled in the browser

### LINKED ISSUE #4070: Consider moving away from assert for code that runs in the browser
The assert library is a native node API that has a client side equivalent that Fluid has been taking a dependency on: 
https://github.com/browserify/commonjs-assert

The library itself is quite small, but it does pull in a lot of dependencies. Specifically it pulls in these dependencies:
```json
    "es6-object-assign": "^1.1.0",
    "is-nan": "^1.2.1",
    "object-is": "^1.0.1",
    "util": "^0.12.0"
```

I did a demo where I modified the webpack config in examples/collaborative-textarea
```javascript
        resolve: {
            extensions: [".ts", ".tsx", ".js"],
            alias: {
                "assert": path.resolve(__dirname, 'assert.js')
            }
        },
```

Where assert.js is a js file in the root of the package with this definition
```javascript
export function strict() {
}
```

This took the bundle down by 7KB gzipped and 24KB uncompressed.

Can we instead write a lightweight isomorphic library that provides the same functionality?

### PR DESCRIPTION
**Pull request by christiango**
Resolves #4070

With this change we remove all client side usages of assert to dramatically reduce bundle sizes. Also adds a custom webpack plugin to the bundle size tests to cause PRs that add new instances of assert will fail to build.

This reduces the container from 240KB (62.5KB compressed) to 180KB (45.5KB compressed)

### CODE REVIEW

**Comment by christiango** [comment]
This PR can't be merged until #4140 and #4139 are merged

File: examples/utils/bundle-size-tests/bannedModulesPlugin.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
+/*!
```
Comment (vladsud) [inline examples/utils/bundle-size-tests/bannedModulesPlugin.js:1]: Does this file need to be here? I.e. should it be separate PR?

File: packages/dds/matrix/src/permutationvector.ts [hunk 2]
```diff
@@ -346,7 +346,7 @@ export class PermutationVector extends Client {
             }
 
             default:
-                assert.fail();
+                throw new Error();
```
Comment (vladsud) [inline packages/dds/matrix/src/permutationvector.ts:349]: Having at least some text here would be nice :)

File: examples/utils/bundle-size-tests/bannedModulesPlugin.js (hunk 1 above)
Comment (christiango) [reply examples/utils/bundle-size-tests/bannedModulesPlugin.js:1]: It keeps it from regressing, I can move it to another PR if that helps review this one.

File: packages/dds/matrix/src/permutationvector.ts (hunk 2 above)
Comment (christiango) [reply packages/dds/matrix/src/permutationvector.ts:349]: Any suggestions for what it should say?

File: packages/dds/ordered-collection/src/consensusOrderedCollection.ts [hunk 3]
```diff
@@ -327,7 +327,9 @@ export class ConsensusOrderedCollection<T = any>
             }
             if (local) {
                 assert(
-                    localOpMetadata, `localOpMetadata is missing from the local client's ${op.opName} operation`);
+                    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
```
Comment (vladsud) [inline packages/dds/ordered-collection/src/consensusOrderedCollection.ts:330]: I'd change cast below to include undefined and move this assert after it, testing resolve for not being undefined.
That said, maybe even better - just remove assert, it does not add much value - we will crash one way or another if things go wrong

File: packages/dds/register-collection/src/consensusRegisterCollection.ts [hunk 4]
```diff
@@ -255,7 +254,10 @@ export class ConsensusRegisterCollection<T>
                         message.sequenceNumber,
                         local);
                     if (local) {
-                        assert(localOpMetadata, "localOpMetadata is missing from the client's write operation");
+                        assert(
```
Comment (vladsud) [inline packages/dds/register-collection/src/consensusRegisterCollection.ts:257]: same here, I'd just remove it

**Review by vladsud — APPROVED** [review]
